Framework/PythonInterface/plugins/algorithms/IntegratePeaksShoeboxTOF.py

```diff
@@ -0,0 +1,687 @@
+                else:
+                    nbins = self.getProperty("NBins").value
+                kernel = make_kernel(nrows, ncols, nbins)
+
+                # get data array and crop
+                peak_data = array_converter.get_peak_data(
+                    peak, detid, bank_name, nshoebox * kernel.shape[0], nshoebox * kernel.shape[1], nrows_edge, ncols_edge
+                )
+                x, y, esq, ispecs = get_and_clip_data_arrays(ws, peak_data, pk_tof, ispec, kernel, nshoebox)
+
+                # perform initial integration
+                intens_over_sig = convolve_shoebox(y, esq, kernel)
+
+                # identify best shoebox position near peak
+                ix = np.argmin(abs(x - pk_tof))
+                ipos = find_nearest_peak_in_data_window(intens_over_sig, ispecs, x, ws, peaks, ipk, peak_data.irow, peak_data.icol, ix)
+
+                # perform final integration if required
+                det_edges = peak_data.det_edges if not integrate_on_edge else None
+                if ipos is not None:
+                    # integrate at that position (without smoothing I/sigma)
+                    intens, sigma, status = integrate_shoebox_at_pos(y, esq, kernel, ipos, weak_peak_threshold, det_edges)
+                    if status == PEAK_STATUS.STRONG and do_optimise_shoebox:
+                        ipos, (nrows, ncols, nbins) = optimise_shoebox(y, esq, (nrows, ncols, nbins), ipos)
+                        kernel = make_kernel(nrows, ncols, nbins)
+                        # re-integrate but this time check for overlap with edge
+                        intens, sigma, status = integrate_shoebox_at_pos(y, esq, kernel, ipos, weak_peak_threshold, det_edges)
+
+                if status == PEAK_STATUS.WEAK and do_optimise_shoebox and weak_peak_strategy == "NearestStrongPeak":
+                    # look for possible strong peaks at any TOF in the window (won't know if strong until all pks integrated)
+                    ipks_near, _ = find_ipks_in_window(ws, peaks, ispecs, ipk)
+                    fwhm = fwhm if get_nbins_from_b2bexp_params else None  # not calculated but not going to be used
+                    weak_peaks_list.append(WeakPeak(ipk, ispecs[ipos[0], ipos[1]], x[ipos[-1]], fwhm, ipks_near))
+                else:
+                    if status == PEAK_STATUS.STRONG:
+                        ipks_strong.append(ipk)
+                    if output_file:
+                        # save result for plotting
+                        i_over_sig = intens / sigma if sigma > 0 else 0.0
+                        results[ipk] = ShoeboxResult(
+                            ipk, peak, x, y, [nrows, ncols, nbins], ipos, [peak_data.irow, peak_data.icol, ix], i_over_sig, status
+                        )
+                # scale summed intensity by bin width to get integrated area
+                intens = intens * bin_width
+                sigma = sigma * bin_width
```

Comment:
As it is, the PyExec function is very long. It is possible to regroup this lines into a different function? Like `optimize_integration_on_peaks` ? or any other name you find descriptive.

Reply:
Yep, it is quite long, I didn't make a new function for precisely the reason you suggest - that it would involve passing a lot of variables. I'll have a look to see if they could be stored in a data class.